gocourse06/homework.go

```diff
@@ -0,0 +1,144 @@
+package main
+
+import (
+	"fmt"
+	"math/rand/v2"
+	"sync"
+	"time"
+)
+
+type Animal struct {
+	Name   string
+	Health int
+	Hunger int
+	State  string
+}
+
+type Cage struct {
+	ID     int
+	IsOpen bool
+}
+
+type Feeder struct {
+	ID     int
+	IsFull bool
+}
+
+func main() {
+	var wg sync.WaitGroup
+	processData(&wg)
+}
+
+func processData(wg *sync.WaitGroup) {
+	animals := makeAnimals()
+	animalsChannel := make(chan Animal)
+	wg.Add(len(animals))
+	for _, animal := range animals {
+		go func(animal Animal) {
+			defer wg.Done()
+			collectAnimalsData(&animal, animalsChannel, time.Duration(rand.IntN(5)))
+		}(animal)
+	}
+
+	cages := makeCages()
+	cagesChannel := make(chan Cage)
+	wg.Add(len(cages))
+	for _, cage := range cages {
+		go func(cage Cage) {
+			defer wg.Done()
+			collectCagesData(&cage, cagesChannel, time.Duration(rand.IntN(5)))
+		}(cage)
+	}
+
+	feeders := makeFeeders()
+	feedersChannel := make(chan Feeder)
+	wg.Add(len(feeders))
+	for _, feeder := range feeders {
+		go func(feeder Feeder) {
+			defer wg.Done()
+			collectFeedersData(feeder, feedersChannel, time.Duration(rand.IntN(5)))
+		}(feeder)
+	}
+
+	go func() {
+		for status := range animalsChannel {
+			fmt.Printf("Received Animal data: %+v\n", status)
+		}
+	}()
+
+	go func() {
+		for cage := range cagesChannel {
+			fmt.Printf("Cage #%d is open %t\n", cage.ID, cage.IsOpen)
+		}
+	}()
+
+	go func() {
+		for feeder := range feedersChannel {
+			fmt.Printf("Feeder #%d is full %t\n", feeder.ID, feeder.IsFull)
+		}
+	}()
```

Comment:
Здається, можна обійтися без горутини:
```suggestion
	for feeder := range feedersChannel {
		fmt.Printf("Feeder #%d is full %t\n", feeder.ID, feeder.IsFull)
	}
```
Горутина - хоч і легенька, але все одно навантажує ресурси. Тому якщо десь можна обійтися без горутини, то краще обходитися.

Пропоную прочитати статтю ["Горутинова спокуса. Приклади використання горутин, де вони насправді зайві"](https://dou.ua/forums/topic/31777/), де наведено інші приклади зайвих горутин.

Reply:
без нього видає помилку
`fatal error: all goroutines are asleep - deadlock!`

Reply:
Дякую, що перевірив.